src/main.cpp

```diff
@@ -2706,10 +2709,16 @@ bool AcceptBlock(CBlock& block, CValidationState& state, CBlockIndex** ppindex,
     if (!AcceptBlockHeader(block, state, &pindex))
         return false;
 
-    if (pindex->nStatus & BLOCK_HAVE_DATA) {
-        // TODO: deal better with duplicate blocks.
-        // return state.DoS(20, error("AcceptBlock(): already have block %d %s", pindex->nHeight, pindex->GetBlockHash().ToString()), REJECT_DUPLICATE, "duplicate");
-        return true;
+    // Try to process all requested blocks that we don't have, but only
+    // process an unrequested block if it's new and has enough work to
+    // advance our tip.
+    bool fAlreadyHave = pindex->nStatus & BLOCK_HAVE_DATA;
+    bool fHasMoreWork = (chainActive.Tip() ? pindex->nChainWork > chainActive.Tip()->nChainWork : true);
```

Comment:
I would either set an error condition (with or without ban), and returnsl
false; or return true. Just logging and returning false may break some
assumptions. You can keep things as is now, but leave the comment about
better dealing with duplicates then.

You're right about nSequenceId, but I would like to only have a single
place in the code for comparing/ordering potential blocks. Anyway, leave it
for now.